### PR TITLE
Add CI for Windows and MacOS platforms

### LINKED ISSUE #1474: Add Windows Support
Coming from https://github.com/opensearch-project/opensearch-plugins/issues/95, add Windows support.

- [x] Passing CI on Windows
- [x] opensearch-project/documentation-website#1714
- [x] https://github.com/opensearch-project/security/issues/1476
- [x] #2232

### PR DESCRIPTION
**Pull request by peternied**
### Description
Add CI for Windows and MacOS platforms

Due to the increase in the number of platforms, I've separated the newer integration tests into their own workflow.  Until retries have been enabled they will automatically pass - but still run and report logs.  As soon as we have full confidence we will allow them to start blocking pull requests from merging.  https://github.com/opensearch-project/security/issues/2184

Switch the gradle commands to be platform agnostic via the `gradle/gradle-build-action@v2`, dropping the 'clean' step to the build which allows us to reuse the gradle cache - if we see any problems pulling in more recent snapshots we can disable this setting quickly.

Found and fixed an issued with config value replacement via environment variables, long story short Windows and MacOS allow for many more characters that are used in the unix environment variable landscape.  Added new tests to cover these interesting scenarios as well.

Found an encoding issue with user names from config files, still unclear of the source of this issue, be it test setup specific or a problem in the broader OpenSearch ecosystem, disabling the `testSpecialUsernames` until we can dive deeper. https://github.com/opensearch-project/security/issues/2194

Disabled the HeapBasedRateTrackerTests - it was depending on system timing and was very brittle if the system under test experienced any undo load, created follow up issue https://github.com/opensearch-project/security/issues/2193

Fixed a test issue in testDlsWithMinDocCountZeroAggregations where there was a random chance for a test failure, easier to find intermittent tests when they are run so often.

OpenSSL has open questions - while it is supported for our Linux JDK11 builds, it seems like a stopgap measure.  I've disabled the tests on windows environment while we determine if we should support OpenSSL at all on Windows JDK11.  https://github.com/opensearch-project/security/issues/2195

### Issues Resolved
- Closes #1474

### Check List
- [X] New functionality includes testing
- [ ] ~New functionality has been documented~
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/security/ssl/OpenSSLTest.java [hunk 1]
```diff
@@ -1,247 +0,0 @@
-/*
```
Comment (scrawfor99) [inline src/test/java/org/opensearch/security/ssl/OpenSSLTest.java:1]: Why was this file deleted with this PR? Was it intentional?

Comment (peternied) [reply src/test/java/org/opensearch/security/ssl/OpenSSLTest.java:1]: These tests don't work on my windows machine, using CI to find the other impacted features/tests.  This will be added back before this is ready for PR or I'll have a good reason to keep it removed

**Comment by peternied** [comment]
Need to look into the following failures:
### MacOS
Tests with failures:
 - [ ] org.opensearch.security.AdvancedSecurityMigrationTests.testWithPassiveAuthDisabled
 - [ ] org.opensearch.security.HttpIntegrationTests.testTenantInfo
 - [ ] org.opensearch.security.IntegrationTests.testUpdate
 - [ ] org.opensearch.security.TransportUserInjectorIntegTest.testSecurityUserInjection
 - [ ] org.opensearch.security.UtilTests.testEnvReplace
 - [ ] org.opensearch.security.auth.limiting.HeapBasedRateTrackerTest.expiryTest

1154 tests completed, 16 failed, 29 skipped
https://github.com/opensearch-project/security/actions/runs/3316861543/jobs/5479123629
 
 ### Windows
Tests with failures:
 - [ ] org.opensearch.security.HttpIntegrationTests.testTenantInfo
 - [ ] org.opensearch.security.IntegrationTests.testSpecialUsernames
 - [ ] org.opensearch.security.SlowIntegrationTests.testDelayInSecurityIndexInitialization
 - [ ] org.opensearch.security.UtilTests.testEnvReplace
 - [ ] org.opensearch.security.auditlog.compliance.ComplianceAuditlogTest.testInternalConfig
 - [ ] org.opensearch.security.auth.limiting.HeapBasedRateTrackerTest.maxTwoTriesTest
 - [ ] org.opensearch.security.dlic.rest.api.TenantInfoActionTest.testTenantInfoAPIUpdate 


1157 tests completed, 18 failed, 29 skipped
https://github.com/opensearch-project/security/actions/runs/3316861543/jobs/5479123554

**Comment by peternied** [comment]
# Latest failures to investigate
https://github.com/opensearch-project/security/actions/runs/3328836236

## JDK11
### Linux
Tests with failures:
 - [ ] org.opensearch.security.TransportUserInjectorIntegTest.testSecurityUserInjection
 - [ ] org.opensearch.security.HttpIntegrationTests.testTenantInfo
 - [ ] org.opensearch.security.dlic.rest.api.legacy.LegacyTenantInfoActionTests.testTenantInfoAPIUpdate

### Windows
_Pending_

### MacOS
Tests with failures:
 - [ ] org.opensearch.security.SlowIntegrationTests.testDelayInSecurityIndexInitialization
 - [ ] org.opensearch.security.ccstest.CrossClusterSearchTests.testCcsWithDiffCertsWithNodesDnStaticallyAdded
 - [ ] org.opensearch.security.dlic.dlsfls.DlsTest.testDlsWithMinDocCountZeroAggregations
 - [ ] org.opensearch.security.dlic.rest.api.legacy.LegacyTenantInfoActionTests.testTenantInfoAPIAccess

## JDK17

### Linux
Tests with failures:
 - [ ] org.opensearch.security.HttpIntegrationTests.testTenantInfo
 - [ ] org.opensearch.security.dlic.dlsfls.DlsTest.testDlsWithMinDocCountZeroAggregations

### Windows
Tests with failures:
 - [ ] org.opensearch.security.HttpIntegrationTests.testTenantInfo
 - [ ] org.opensearch.security.TransportUserInjectorIntegTest.testSecurityUserInjection
 - [ ] org.opensearch.security.dlic.dlsfls.DlsTest.testDlsWithMinDocCountZeroAggregations
 - [ ] org.opensearch.security.dlic.rest.api.legacy.LegacyTenantInfoActionTests.testTenantInfoAPIAccess
 - [ ] org.opensearch.security.dlic.rest.api.legacy.LegacyWhitelistApiTests.testWhitelistInvalidHttpRequestMethod

### MacOS
Tests with failures:
 - [ ] org.opensearch.security.SystemIntegratorsTests.testInjectedUserMalformed
 - [ ] org.opensearch.security.dlic.dlsfls.DlsTest.testDlsWithMinDocCountZeroAggregations
 - [ ] org.opensearch.security.dlic.rest.api.TenantInfoActionTest.testTenantInfoAPIUpdate

File: src/test/java/org/opensearch/security/dlic/dlsfls/DlsTest.java [hunk 2]
```diff
@@ -366,7 +366,7 @@ public void testDlsWithMinDocCountZeroAggregations() throws Exception {
 
         // Significant Text Aggregation is not impacted.
         // Non-admin user with setting "min_doc_count=0". Expected to only have access to buckets for dept_manager".
-        String query3 = "{\"aggregations\":{\"significant_termX\":{\"significant_terms\":{\"field\":\"termX.keyword\",\"min_doc_count\":0}}}}";
+        String query3 = "{\"size\":100,\"aggregations\":{\"significant_termX\":{\"significant_terms\":{\"field\":\"termX.keyword\",\"min_doc_count\":0}}}}";
```
Comment (peternied) [inline src/test/java/org/opensearch/security/dlic/dlsfls/DlsTest.java:369]: This test inserts 17 documents and then does a query for matches on them checking the 'hits'.  This hits have a default limit of 10... so this test was failing all the time because it depends if you got the right mix of 10/17 items in the output.  By making this size much larger this should prevent this issue from happening again.

Always be careful with defaults!

File: src/test/java/org/opensearch/security/test/helper/file/FileHelper.java [hunk 3]
```diff
@@ -108,6 +109,8 @@ public static BytesReference readYamlContent(final String file) {
             parser.nextToken();
             final XContentBuilder builder = XContentFactory.jsonBuilder();
             builder.copyCurrentStructure(parser);
+            // Reproduction
```
Comment (peternied) [inline src/test/java/org/opensearch/security/test/helper/file/FileHelper.java:112]: Need to remove, this is where I found #2194

File: src/test/java/org/opensearch/security/ssl/OpenSSLTest.java [hunk 4]
```diff
@@ -22,7 +22,6 @@
 import java.util.Set;
```
Comment (peternied) [inline src/test/java/org/opensearch/security/ssl/OpenSSLTest.java:22]: These tests were widely untested, waiting for the latest test run to confirm if this updated behaviors works correctly in linux jdk11 runs

This is one of the last items to remove the 'Draft' state of this pull request

File: src/test/java/org/opensearch/security/ssl/OpenSSLTest.java (hunk 1 above)
Comment (peternied) [reply src/test/java/org/opensearch/security/ssl/OpenSSLTest.java]: This has been resolved - please re-review around here

File: src/test/java/org/opensearch/security/ssl/OpenSSLTest.java (hunk 4 above)
Comment (peternied) [reply src/test/java/org/opensearch/security/ssl/OpenSSLTest.java]: Resolved!

**Comment by peternied** [comment]
@opensearch-project/security This change is ready for review.  Note; I assuming this change is approved I'll adjust the branch protection rules for the new matrix of builds

**Review by davidlago — APPROVED** [review]

**Comment by peternied** [comment]
Manually backporting in #2205
